docs/libcudacxx/extended_api/execution_model.rst

```diff
@@ -0,0 +1,261 @@
+
+The order in which device threads make progress is _unspecified_.
+
+Modify `[intro.progress.1] <https://eel.is/c++draft/intro.progress>`__ as follows (modifications in **bold**):
+
+The implementation may assume that any **host** thread will eventually do one of the following:
+
+    1. terminate,
+    2. invoke the function `std::this_thread::yield <https://en.cppreference.com/w/cpp/thread/yield>`__ (`[thread.thread.this] <http://eel.is/c++draft/thread.thread.this>`__),
+    3. make a call to a library I/O function,
+    4. perform an access through a volatile glvalue,
+    5. perform a synchronization operation or an atomic operation, or
+    6. continue execution of a trivial infinite loop (`[stmt.iter.general] <http://eel.is/c++draft/stmt.iter.general>`__).
+
+**The implementation may assume that any device thread will eventually do one of the following:**
+
+    1. **terminate**,
+    2. **make a call to a library I/O function**,
+    3. **perform an access through a volatile glvalue except if the designated object has automatic storage duration, or**
+    4. **perform a synchronization operation or an atomic read operation except if the designated object has automatic storage duration.**
+
+  [Note: Some current limitations of device threads relative to host threads are considered
+  implementation defects, such as the undefined behavior that arises from device threads
+  that eventually only perform volatile or atomic operations on automatic storage duration objects.
+  However, other limitations stem from the strictness of the C++ standard, e.g., providing
+  forward progress to programs that eventually only perform atomic writes or fences degrades overall
+  performance for little practical benefit. - end note.]
+
+.. dropdown:: Examples of forward progress guarantee differences between host and device threads due to modifications to [intro.progress.1].
+
+    The following examples refer to the itemized sub-clauses of the implementation assumptions for host and device threads above
+    using "host.threads.<id>" and "device.threads.<id>", respectively.
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.0
+        // Outcome: grid eventually terminates per device.threads.4 because the atomic object does not have automatic storage duration.
+        __global__ void ex0(cuda::atomic_ref<int, cuda::thread_scope_device> atom) {
+            if (threadIdx.x == 0) {
+                while(atom.load(cuda::memory_order_relaxed) == 0);
+            } else if (threadIdx.x == 1) {
+                atom.store(1, cuda::memory_order_relaxed);
+            }
+        }
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.1
+        // Allowed outcome: No thread makes progress because device threads don't support host.threads.2.
+        __global__ void ex1() {
+            while(true) std::this_thread::yield();
+        }
```

Comment:
Would this even compile?  Perhaps a better question to ask, is how would I spell the equivalent idea in device code?

Reply:
```suggestion
        // Example: Execution.Model.Device.1
        // Allowed outcome: No thread makes progress because device threads don't support host.threads.2.
        __global__ void ex1() {
            while(true) cuda::std::this_thread::yield();
        }
```